Add gramine-manifest-check to GSC build

DukeDavis12 opened the pull request:
<!--
    Please fill in the following form before submitting this PR
    and ensure that your code follows our coding style guideline:
    https://gramine.readthedocs.io/en/latest/devel/coding-style.html -->

## Description of the changes <!-- (reasons and measures) -->
The gramine-manifest-check tool for manifest schema checks was introduced in the core Gramine repository.

This commit integrates gramine-manifest-check tool into the GSC build template and updates related
manifest templates.
      
Fixes issue #219.

<!--
    If your PR fixes an issue, please remember to add "Fixes #issue_number"
    here, to automatically close it on merge. -->

## How to test this PR? <!-- (if applicable) -->

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/gramineproject/gsc/223)
<!-- Reviewable:end -->

woju reviewed (changes requested):
Did you test this, or you just edited, committed, and sent a PR without even running it a single time?

> *[Reviewable](https://reviewable.io/reviews/gramineproject/gsc/223#-:-OB-xv7D92UJgxWkDT4z:b5482df)* status: 0 of 1 files reviewed, all discussions resolved, not enough approvals from maintainers (3 more required), not enough approvals from different teams (2 more required, approved so far: ), one-liner is not in imperative mood



<!-- Sent from Reviewable.io -->

DukeDavis12 commented:
@woju  Didn't you see the draft status on PR ? 

Testing is on progress; Plus, I am ramping on Gramine / TDX. 

Will mark the PR "Ready for review" once one complete round testing is done.

DukeDavis12 commented:
Fails if user provides deprecated sgx properties
![image](https://github.com/user-attachments/assets/e661172d-ef95-4eca-8340-ea224fc34927)

Pass case
![image](https://github.com/user-attachments/assets/e545ac5b-5d60-4fe9-abcc-27d5c8d93a21)

mkow commented:
@DukeDavis12: Please don't paste screenshots of text...

@woju: Could you review this now?

woju reviewed (changes requested):
> Reviewed 10 of 10 files at r2.
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gsc/223)* status: all files reviewed, 4 unresolved discussions, not enough approvals from maintainers (2 more required), not enough approvals from different teams (2 more required, approved so far: ), one-liner is not in imperative mood (waiting on @DukeDavis12)

___
*[`templates/entrypoint.common.manifest.template` line 15 at r2](https://reviewable.io/reviews/gramineproject/gsc/223#-OC-ITmZ7Y0kEXJ5KHPI:-OC-ITmZ7Y0kEXJ5KHPJ:bw7wmzd) ([raw file](https://github.com/gramineproject/gsc/blob/c72dcdf7d5584249db5bbb826bfa0b99f4f6d8ae/templates/entrypoint.common.manifest.template#L15)):*
> ```
> fs.start_dir = "{{working_dir}}"
> fs.mounts = [
> ]
> ```

Why this is needed?
___
*[`templates/debian/Dockerfile.build.template` line 54 at r2](https://reviewable.io/reviews/gramineproject/gsc/223#-OC-I2Ds4LjK9DaD3vGN:-OC-I2Ds4LjK9DaD3vGO:bqkdgad) ([raw file](https://github.com/gramineproject/gsc/blob/c72dcdf7d5584249db5bbb826bfa0b99f4f6d8ae/templates/debian/Dockerfile.build.template#L54)):*
> ```
> {% endblock %}
> 
> {% block path %}export PYTHONPATH="${PYTHONPATH}:$(find /gramine/meson_build_output/lib -type d -path '*/site-packages')" &&{% endblock %}
> ```

I feel like there's too much duplication of this PATH, between different distros and also between different templates (build vs sign), which you've surely saw (I suppose that's why you moved this to build and not sign). Isn't there any way to specify this only twice in the whole project, once for debian-based distros and once for rhel-based?
___
*[`templates/centos/Dockerfile.build.template` line 42 at r2](https://reviewable.io/reviews/gramineproject/gsc/223#-OC-HO5G9oj7Nau_pQ5f:-OC-HO5HAEBADQ_pBeat:bflnxat) ([raw file](https://github.com/gramineproject/gsc/blob/c72dcdf7d5584249db5bbb826bfa0b99f4f6d8ae/templates/centos/Dockerfile.build.template#L42)):*
> ```
> {% endblock %}
> 
> {% block path %}export PYTHONPATH="${PYTHONPATH}:$(find /gramine/meson_build_output/lib64 -type d -path '*/site-packages')" &&{% endblock %}
> ```

no newline at EOF
___
*[`templates/redhat/ubi-minimal/Dockerfile.build.template` line 50 at r2](https://reviewable.io/reviews/gramineproject/gsc/223#-OB_eQ83DrKfyoQAwlkk:-OB_eQ845B_ejTeYnxWi:bflnxat) ([raw file](https://github.com/gramineproject/gsc/blob/c72dcdf7d5584249db5bbb826bfa0b99f4f6d8ae/templates/redhat/ubi-minimal/Dockerfile.build.template#L50)):*
> ```
> {% endblock %}
> 
> {% block path %}export PYTHONPATH="${PYTHONPATH}:$(find /gramine/meson_build_output/lib64 -type d -path '*/site-packages')" &&{% endblock %}
> ```

no newline at EOF


<!-- Sent from Reviewable.io -->

DukeDavis12 commented:
> > Reviewed 10 of 10 files at r2.
> > _[Reviewable](https://reviewable.io/reviews/gramineproject/gsc/223)_ status: all files reviewed, 4 unresolved discussions, not enough approvals from maintainers (2 more required), not enough approvals from different teams (2 more required, approved so far: ), one-liner is not in imperative mood (waiting on @DukeDavis12)
> 
> _[`templates/entrypoint.common.manifest.template` line 15 at r2](https://reviewable.io/reviews/gramineproject/gsc/223#-OC-ITmZ7Y0kEXJ5KHPI:-OC-ITmZ7Y0kEXJ5KHPJ:bw7wmzd) ([raw file](https://github.com/gramineproject/gsc/blob/c72dcdf7d5584249db5bbb826bfa0b99f4f6d8ae/templates/entrypoint.common.manifest.template#L15)):_
> 
> > ```
> > fs.start_dir = "{{working_dir}}"
> > fs.mounts = [
> > ]
> > ```
> 
> Why this is needed?

fs.mounts option is made mandatory gramine-manifest-check tool. https://github.com/gramineproject/gramine/blob/master/python/graminelibos/manifest_check.py#L49

> 
> _[`templates/debian/Dockerfile.build.template` line 54 at r2](https://reviewable.io/reviews/gramineproject/gsc/223#-OC-I2Ds4LjK9DaD3vGN:-OC-I2Ds4LjK9DaD3vGO:bqkdgad) ([raw file](https://github.com/gramineproject/gsc/blob/c72dcdf7d5584249db5bbb826bfa0b99f4f6d8ae/templates/debian/Dockerfile.build.template#L54)):_
> 
> > ```
> > {% endblock %}
> > 
> > {% block path %}export PYTHONPATH="${PYTHONPATH}:$(find /gramine/meson_build_output/lib -type d -path '*/site-packages')" &&{% endblock %}
> > ```
> 
> I feel like there's too much duplication of this PATH, between different distros and also between different templates (build vs sign), which you've surely saw (I suppose that's why you moved this to build and not sign). Isn't there any way to specify this only twice in the whole project, once for debian-based distros and once for rhel-based?
> 
> _[`templates/centos/Dockerfile.build.template` line 42 at r2](https://reviewable.io/reviews/gramineproject/gsc/223#-OC-HO5G9oj7Nau_pQ5f:-OC-HO5HAEBADQ_pBeat:bflnxat) ([raw file](https://github.com/gramineproject/gsc/blob/c72dcdf7d5584249db5bbb826bfa0b99f4f6d8ae/templates/centos/Dockerfile.build.template#L42)):_
> 
> > ```
> > {% endblock %}
> > 
> > {% block path %}export PYTHONPATH="${PYTHONPATH}:$(find /gramine/meson_build_output/lib64 -type d -path '*/site-packages')" &&{% endblock %}
> > ```
> 
> no newline at EOF
Updated.

> 
> _[`templates/redhat/ubi-minimal/Dockerfile.build.template` line 50 at r2](https://reviewable.io/reviews/gramineproject/gsc/223#-OB_eQ83DrKfyoQAwlkk:-OB_eQ845B_ejTeYnxWi:bflnxat) ([raw file](https://github.com/gramineproject/gsc/blob/c72dcdf7d5584249db5bbb826bfa0b99f4f6d8ae/templates/redhat/ubi-minimal/Dockerfile.build.template#L50)):_
> 
> > ```
> > {% endblock %}
> > 
> > {% block path %}export PYTHONPATH="${PYTHONPATH}:$(find /gramine/meson_build_output/lib64 -type d -path '*/site-packages')" &&{% endblock %}
> > ```
> 
> no newline at EOF
Updated.

chiache reviewed (changes requested):
> Reviewed 1 of 10 files at r2, 3 of 4 files at r5, 1 of 2 files at r6, 2 of 2 files at r7.
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gsc/223)* status: 10 of 11 files reviewed, 7 unresolved discussions, not enough approvals from maintainers (2 more required), not enough approvals from different teams (2 more required, approved so far: ), one-liner is not in imperative mood (waiting on @DukeDavis12 and @woju)

___
*[`finalize_manifest.py` line 161 at r7](https://reviewable.io/reviews/gramineproject/gsc/223#-OF8UIxyBIWBYTYwZbFO:-OF8UIxyBIWBYTYwZbFP:bpe33qz) ([raw file](https://github.com/gramineproject/gsc/blob/1c563227497b172acbeaee205bd71fa276543838/finalize_manifest.py#L161)):*
> ```Python
>     if 'uri' not in rendered_manifest_dict['loader']['entrypoint']:
>         rendered_manifest_dict['loader']['entrypoint']['uri'] = "file:/gramine/meson_build_output/lib/x86_64-linux-gnu/gramine/libsysdb.so"
> ```

Why not add `/gramine/meson_build_output/lib64/gramine/runtime/glibc` to `loader.env.LD_LIBRARY_PATH` in the script?
___
*[`templates/debian/entrypoint.manifest.template` line 7 at r7](https://reviewable.io/reviews/gramineproject/gsc/223#-OF8UXT20i7_Ss7Kxn9O:-OF8UXT20i7_Ss7Kxn9P:b-7ropln) ([raw file](https://github.com/gramineproject/gsc/blob/1c563227497b172acbeaee205bd71fa276543838/templates/debian/entrypoint.manifest.template#L7)):*
> ```
> # Add "/usr/lib/x86_64-linux-gnu" explicitly because ldconfig in Ubuntu 21.04 doesn't
> # produce it; note that this Debian template is used by Ubuntu templates as well
> loader.env.LD_LIBRARY_PATH = "/gramine/meson_build_output/lib/x86_64-linux-gnu/gramine/runtime/glibc:/usr/lib/x86_64-linux-gnu:{{"{{library_paths}}"}}"
> ```

ditto
___
*[`templates/centos/entrypoint.manifest.template` line 5 at r7](https://reviewable.io/reviews/gramineproject/gsc/223#-OF8UVXQAXgw6DnFXKeb:-OF8UVXQAXgw6DnFXKec:b-7ropln) ([raw file](https://github.com/gramineproject/gsc/blob/1c563227497b172acbeaee205bd71fa276543838/templates/centos/entrypoint.manifest.template#L5)):*
> ```
> {% block loader %}
> loader.env.LD_LIBRARY_PATH = "/gramine/meson_build_output/lib64/gramine/runtime/glibc:/usr/lib64:{{"{{library_paths}}"}}"
> {% endblock %}
> ```

ditto


<!-- Sent from Reviewable.io -->

DukeDavis12 commented:
@chiache Will create a separate PR for removing loader.entrypoint.uri from manifest file.

kailun-qin reviewed (changes requested):
Could you pls create a separate PR that only contains the fix for https://github.com/gramineproject/gsc/issues/224 and closes https://github.com/gramineproject/gsc/pull/225?

> Reviewed 1 of 4 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gsc/223#-:-OFulC5QCaEyVh7sLlkh:b-b6b0i)* status: 7 of 11 files reviewed, 8 unresolved discussions, not enough approvals from maintainers (2 more required), not enough approvals from different teams (2 more required, approved so far: ), one-liner is not in imperative mood (waiting on @chiache, @DukeDavis12, and @woju)

___
*[`finalize_manifest.py` line 28 at r8](https://reviewable.io/reviews/gramineproject/gsc/223#-OFusTtSB6D7EZx7rRPa:-OFusTtSB6D7EZx7rRPb:bvfar8a) ([raw file](https://github.com/gramineproject/gsc/blob/163bbb7d7ace7ad0b63f012f29c7b7f0cbe1fccf/finalize_manifest.py#L28)):*
> ```Python
>     sha256_hash = hashlib.sha256()
>     with open(filename, 'rb') as f:
>         for byte_block in iter(lambda: f.read(4096), b""):
> ```

pls don't hardcode, you can do sth like: https://github.com/gramineproject/gramine/blob/21ae038ba1ddd2c8f00e064acb5b727fc92af0c7/python/graminelibos/manifest.py#L230.



_Code quote:_
```Python
4096
```


<!-- Sent from Reviewable.io -->

DukeDavis12 commented:
> Could you pls create a separate PR that only contains the fix for #224 and closes #225?
- Sure WIP.
> 
> > Reviewed 1 of 4 files at r8, all commit messages.
> > _[Reviewable](https://reviewable.io/reviews/gramineproject/gsc/223#-:-OFulC5QCaEyVh7sLlkh:b-b6b0i)_ status: 7 of 11 files reviewed, 8 unresolved discussions, not enough approvals from maintainers (2 more required), not enough approvals from different teams (2 more required, approved so far: ), one-liner is not in imperative mood (waiting on @chiache, @DukeDavis12, and @woju)
> 
> _[`finalize_manifest.py` line 28 at r8](https://reviewable.io/reviews/gramineproject/gsc/223#-OFusTtSB6D7EZx7rRPa:-OFusTtSB6D7EZx7rRPb:bvfar8a) ([raw file](https://github.com/gramineproject/gsc/blob/163bbb7d7ace7ad0b63f012f29c7b7f0cbe1fccf/finalize_manifest.py#L28)):_
> 
> > ```python
> >     sha256_hash = hashlib.sha256()
> >     with open(filename, 'rb') as f:
> >         for byte_block in iter(lambda: f.read(4096), b""):
> > ```
> 
> pls don't hardcode, you can do sth like: https://github.com/gramineproject/gramine/blob/21ae038ba1ddd2c8f00e064acb5b727fc92af0c7/python/graminelibos/manifest.py#L230.
> 
> _Code quote:_
> 
> ```python
> 4096
> ```

- Will update it in the next PR.

kailun-qin reviewed (changes requested):
> Reviewed 3 of 10 files at r2, 1 of 2 files at r6, 1 of 2 files at r7, 3 of 4 files at r8, 3 of 3 files at r9, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gsc/223)* status: all files reviewed, 9 unresolved discussions, not enough approvals from maintainers (2 more required), not enough approvals from different teams (2 more required, approved so far: ), one-liner is not in imperative mood, "Merge branch '" found in commit messages' one-liners (waiting on @DukeDavis12 and @woju)

___
*[`templates/centos/entrypoint.manifest.template` line 4 at r9](https://reviewable.io/reviews/gramineproject/gsc/223#-OG3mkXwCTjJSssP62h0:-OG3mkXwCTjJSssP62h1:b-kecdxj) ([raw file](https://github.com/gramineproject/gsc/blob/4fa8677fcf93a8cc17111763a9b8af1da489bffa/templates/centos/entrypoint.manifest.template#L4)):*
> ```
> 
> {% block loader %}
> loader.entrypoint.uri = "file:/gramine/meson_build_output/lib64/gramine/libsysdb.so"
> ```

`gramine-direct` is still broken w/ this change, right? (also mentioned in https://github.com/gramineproject/gsc/pull/225#issuecomment-2482039302)



_Code quote:_
```
loader.entrypoint.uri = "file:/gramine/meson_build_output/lib64/gramine/libsysdb.so"
```
___
*[`templates/entrypoint.common.manifest.template` line 3 at r9](https://reviewable.io/reviews/gramineproject/gsc/223#-OG3kwp86kM5LzZfm8XR:-OG3kwp86kM5LzZfm8XS:b-86ix3j) ([raw file](https://github.com/gramineproject/gsc/blob/4fa8677fcf93a8cc17111763a9b8af1da489bffa/templates/entrypoint.common.manifest.template#L3)):*
> ```
> libos.entrypoint = "/gramine/app_files/{{binary_basename}}"
> 
> # Add distro-specific `loader.entrypoint` and `loader.env.LD_LIBRARY_PATH`
> ```

-> `loader.entrypoint.uri`, to align w/ your changes



_Code quote:_
```
loader.entrypoint
```


<!-- Sent from Reviewable.io -->

kailun-qin reviewed (approved):
> Reviewed 1 of 1 files at r10, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gsc/223)* status: all files reviewed, 8 unresolved discussions, not enough approvals from maintainers (1 more required), not enough approvals from different teams (1 more required, approved so far: Intel) (waiting on @DukeDavis12 and @woju)

___
*[`-- commits` line 1 at r10](https://reviewable.io/reviews/gramineproject/gsc/223#-OG45K3n8HnjENte84B3:-OG45K3o-GJai-nVCWDv:b-e3x162):*
Pls consolidate the commits. Can be done during rebase.


<!-- Sent from Reviewable.io -->

DukeDavis12 commented:
@woju  Can you review ?

aneessahib commented:
@woju  - Could you please review this? SGX is broken without merging this PR.

woju reviewed (commented):
@DukeDavis12 @aneessahib I've clicked through most of the stuff, but ISTM you forgot to answer one of the questions.

> Reviewed 1 of 2 files at r6, 1 of 2 files at r7, 2 of 4 files at r8, 3 of 3 files at r9, 1 of 1 files at r10, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gsc/223#-:-OIA2fJ0AGPIudEk1rNY:b-sd9ziv)* status: all files reviewed, 5 unresolved discussions, not enough approvals from maintainers (1 more required), not enough approvals from different teams (1 more required, approved so far: Intel) (waiting on @DukeDavis12)



<!-- Sent from Reviewable.io -->

DukeDavis12 commented:
> @DukeDavis12 @aneessahib I've clicked through most of the stuff, but ISTM you forgot to answer one of the questions.
> 
> > Reviewed 1 of 2 files at r6, 1 of 2 files at r7, 2 of 4 files at r8, 3 of 3 files at r9, 1 of 1 files at r10, all commit messages.
> > _[Reviewable](https://reviewable.io/reviews/gramineproject/gsc/223#-:-OIA2fJ0AGPIudEk1rNY:b-sd9ziv)_ status: all files reviewed, 5 unresolved discussions, not enough approvals from maintainers (1 more required), not enough approvals from different teams (1 more required, approved so far: Intel) (waiting on @DukeDavis12)

Hi Woju,
Moved Deb & RHEL python path to a common location and we are now referencing it.

DukeDavis12 commented on templates/python.common.path.template:
```diff
@@ -0,0 +1,4 @@
+{%- macro python_path(distro) -%}
+{% if 'debian' in distro or 'ubuntu' in distro %}export PYTHONPATH="${PYTHONPATH}:$(find /gramine/meson_build_output/lib -type d -path '*/site-packages')" {% else %}export PYTHONPATH="${PYTHONPATH}:$(find /gramine/meson_build_output/lib64 -type d -path '*/site-packages')"{% endif %}
```
- Written in a single line to avoid introducing unwanted new lines in the final rendered files. 
- Any whitespace, including new lines, within the template tags will be preserved in the output.
- Tried adding `\`; but this introduces unwanted new lines in apploader.template file and sign.template file.

By keeping the logic in a single line, we ensure that the rendered output remains clean and does not contain unnecessary line breaks.

DukeDavis12 commented:
@woju  Updated PR with python path macro.

DukeDavis12 commented on templates/python.common.path.template:
```diff
@@ -0,0 +1,3 @@
+{%- macro python_path() -%}
+{% if 'debian' in Distro or 'ubuntu' in Distro %}export PYTHONPATH="${PYTHONPATH}:$(find /gramine/meson_build_output/lib -type d -path '*/site-packages')" {% else %}export PYTHONPATH="${PYTHONPATH}:$(find /gramine/meson_build_output/lib64 -type d -path '*/site-packages')"{% endif %}
```
- Written in a single line to avoid introducing unwanted new lines in the final rendered files. 
- Any whitespace, including new lines, within the macro tags will be preserved in the output.
- Tried adding `\`; but this introduces unwanted new lines in apploader.template file and sign.template file.

By keeping the logic in a single line, we ensure that the rendered output remains clean and does not contain unnecessary line breaks.

woju replied:
1) Line breaks can be skipped by adding `-` as part of opening or closing tags: `{%-` vs `{%`, `-%}` vs `%}` etc. (also works for `{{-`, `-}}`, `{#-`, `-#}`). Note I did include them in the example. (https://jinja.palletsprojects.com/en/stable/templates/#whitespace-control — also `+` is described there, which is almost certainly not what we want).
2) I don't like that you repeat the whole line instead of making a difference only in the part that actually differ (the path). This way you gain nothing really, readability is similarly pool.
3) `{% if '...' in Distro %}` is wrong, because you need to hardcode the distros in yet another place. What happened if we added another Debian-like distro (say Mint, although it's a desktop-oriented distribution, so probably there's no reason to add that particular one)?

Seriously, the approach with an argument to the macro was given to you for a reason.
<!-- Reviewable comment -OJJUfEh2mzXykBU6hku:b-o3zpns -->
<!-- Sent from Reviewable.io -->

jkr0103 replied:
> `{% if '...' in Distro %}` is wrong, because you need to hardcode the distros in yet another place. What happened if we added another Debian-like distro (say Mint, although it's a desktop-oriented distribution, so probably there's no reason to add that particular one)?


@woju, the statement `you need to hardcode the distros in yet another place` is not correct. `Distro` is detected either from [input image](https://github.com/gramineproject/gsc/blob/master/gsc.py#L320) or from [config.yaml](https://github.com/gramineproject/gsc/blob/master/config.yaml.template#L14).

Hence I think it's safe to use the already available variable `Distro` here.
<!-- Reviewable comment -OJMGnPM5M5HSzUn0fDa:b-dvz5yn -->
<!-- Sent from Reviewable.io -->

woju replied:
This is backwards. The previous version very much did hardcode the fact that only for certain distros some logic happens. Logic should be concentrated in distro-specific subdirectories of templates. But it happened to also be in some other place — which is wrong and what I called hardcoding. Now it's OK.
<!-- Reviewable comment -OJra5714BAqIg-bv530:bb1my5m -->
<!-- Sent from Reviewable.io -->

jkr0103 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gsc/223)* status: 3 of 25 files reviewed, 8 unresolved discussions, not enough approvals from maintainers (2 more required), not enough approvals from different teams (2 more required, approved so far: ) (waiting on @chiache, @DukeDavis12, @kailun-qin, and @woju)

___
*[`templates/apploader.common.template` line 14 at r13](https://reviewable.io/reviews/gramineproject/gsc/223#-OJIJ5YMAkB_WKeDVMJf:-OJIJ5YMAkB_WKeDVMJg:bdcpyd6) ([raw file](https://github.com/gramineproject/gsc/blob/96aa3b94094930d8705938b2437f9a96eb15d4dd/templates/apploader.common.template#L14)):*
> ```
> {{ python_path() }}
> 
> {% block path %}{% endblock %}
> ```

Suggestion: I feel `{% block path %}{% endblock %}` can be replaced completely with below code snippet with `PKG_CONFIG_PATH` definition here itself instead of having it's definition at multiple places. Not blocking on it.



_Code snippet:_
```
{% if 'debian' in Distro or 'ubuntu' in Distro %}
export PKG_CONFIG_PATH="${PKG_CONFIG_PATH}:$(find /gramine/meson_build_output/lib -type d -path '*/pkgconfig')"
{% else %}
export PKG_CONFIG_PATH="${PKG_CONFIG_PATH}:$(find /gramine/meson_build_output/lib64 -type d -path '*/pkgconfig')"
{% endif %}
```


<!-- Sent from Reviewable.io -->

jkr0103 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gsc/223)* status: 3 of 25 files reviewed, 8 unresolved discussions, not enough approvals from maintainers (2 more required), not enough approvals from different teams (2 more required, approved so far: ) (waiting on @chiache, @DukeDavis12, @kailun-qin, and @woju)

___
*[`templates/Dockerfile.common.build.template` line 66 at r13](https://reviewable.io/reviews/gramineproject/gsc/223#-OJIM2pP1OHJv0hBW431:-OJIM2pP1OHJv0hBW432:bry4sts) ([raw file](https://github.com/gramineproject/gsc/blob/96aa3b94094930d8705938b2437f9a96eb15d4dd/templates/Dockerfile.common.build.template#L66)):*
> ```
>     && rm -f /gramine/app_files/finalize_manifest.py
> 
> RUN {{ python_path() }} && gramine-manifest-check /gramine/app_files/entrypoint.manifest
> ```

Can you split in two lines, this line become long once `python_path` is repalced.
___
*[`templates/Dockerfile.common.sign.template` line 15 at r13](https://reviewable.io/reviews/gramineproject/gsc/223#-OJIMKVl-G4V8U-Ob1af:-OJIMKVl-G4V8U-Ob1ag:b-7ropln) ([raw file](https://github.com/gramineproject/gsc/blob/96aa3b94094930d8705938b2437f9a96eb15d4dd/templates/Dockerfile.common.sign.template#L15)):*
> ```
> ARG passphrase
> 
> RUN {{ python_path() }} && gramine-sgx-sign \
> ```

ditto


<!-- Sent from Reviewable.io -->

woju reviewed (changes requested):
> Reviewed 6 of 16 files at r11, 9 of 15 files at r12, 5 of 7 files at r13, 2 of 2 files at r14, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gsc/223)* status: all files reviewed, 6 unresolved discussions, not enough approvals from maintainers (2 more required), not enough approvals from different teams (2 more required, approved so far: ), one-liner is not in imperative mood (waiting on @DukeDavis12)

___
*[`gsc.py` line 395 at r11](https://reviewable.io/reviews/gramineproject/gsc/223#-OIp_f8hBj-fm5rICD9Z:-OJJWZGZ7p8JmXYHeQGo:b-2hhjyt) ([raw file](https://github.com/gramineproject/gsc/blob/c1088f5b2e8695f3be7a0a4380adfe15dd558ff8/gsc.py#L395)):*
<details><summary><i>Previously, woju (Wojtek Porczyk) wrote…</i></summary><blockquote>

This does not make any sense. There's nothing wrong in harcoding the name of the macro (e.g. `{% include 'python_path' %}` — note quotes). Second, this makes you repeat `{% extends %}`, whereas if you hardcoded the path, you wouldn't have to have all those `{% extends %}`.

More important, the files for debian and RHEL are very similar and you need to look carefully to notice `lib` vs `lib64`.

Can you try with a macro?

```jinja2
{%- macro python_path(libdir) -%}
env PYTHONPATH="${PYTHONPATH}:$(find /gramine/meson_build_output{{ libdir }} -type d -path '*/site-packages')"
{%- endmacro -%}
```

```jinja2
{% from 'python_path' import python_path -%}

RUN {{ python_path('/lib') }} command1
RUN {{ python_path('/lib64') }} command2
```
</blockquote></details>

Closing this and one more thread, will continue in new one, near the macro.


<!-- Sent from Reviewable.io -->

jkr0103 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gsc/223)* status: all files reviewed, 6 unresolved discussions, not enough approvals from maintainers (2 more required), not enough approvals from different teams (2 more required, approved so far: ), one-liner is not in imperative mood (waiting on @DukeDavis12 and @woju)



<!-- Sent from Reviewable.io -->

DukeDavis12 commented on gsc.py:
```diff
@@ -572,6 +575,8 @@ def gsc_sign_image(args):
     sign_template = env.get_template(f'{template_path(distro)}/Dockerfile.sign.template')
+    python_path = env.get_template('python.common.path.template')
+    env.globals.update({'python_path': python_path.module.python_path})
     os.makedirs(tmp_build_path, exist_ok=True)
```
loading macro for signing phase.

woju replied:
ditto (fine)
<!-- Reviewable comment -OJrakN82L0WIDT9RC1x:bldnwki -->
<!-- Sent from Reviewable.io -->

DukeDavis12 commented on gsc.py:
```diff
@@ -390,6 +390,9 @@ def gsc_build(args):
+    python_path = env.get_template('python.common.path.template')
+    env.globals.update({'python_path': python_path.module.python_path})
+
     # generate Dockerfile.build from Jinja-style templates/<distro>/Dockerfile.build.template
```
loading macro for building phase.

woju replied:
OK, that also works. I guess it's fine.
<!-- Reviewable comment -OJraecI0bUwgrYkAO2R:bw4jife -->
<!-- Sent from Reviewable.io -->

DukeDavis12 commented:
> > Reviewed 6 of 16 files at r11, 9 of 15 files at r12, 5 of 7 files at r13, 2 of 2 files at r14, all commit messages.
> > _[Reviewable](https://reviewable.io/reviews/gramineproject/gsc/223)_ status: all files reviewed, 6 unresolved discussions, not enough approvals from maintainers (2 more required), not enough approvals from different teams (2 more required, approved so far: ), one-liner is not in imperative mood (waiting on @DukeDavis12)
> 
> _[`gsc.py` line 395 at r11](https://reviewable.io/reviews/gramineproject/gsc/223#-OIp_f8hBj-fm5rICD9Z:-OJJWZGZ7p8JmXYHeQGo:b-2hhjyt) ([raw file](https://github.com/gramineproject/gsc/blob/c1088f5b2e8695f3be7a0a4380adfe15dd558ff8/gsc.py#L395)):_
> 
> Previously, woju (Wojtek Porczyk) wrote…
> Closing this and one more thread, will continue in new one, near the macro.

@woju  Updated PR with macro. Please review.

woju reviewed (approved):
> Reviewed 17 of 17 files at r15, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gsc/223)* status: all files reviewed, 4 unresolved discussions, not enough approvals from maintainers (1 more required), not enough approvals from different teams (1 more required, approved so far: ITL), one-liner is not in imperative mood (waiting on @DukeDavis12)



<!-- Sent from Reviewable.io -->

kailun-qin reviewed (approved):
> Reviewed 16 of 17 files at r15, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gsc/223)* status: all files reviewed, 4 unresolved discussions, one-liner is not in imperative mood (waiting on @DukeDavis12)



<!-- Sent from Reviewable.io -->

jkr0103 reviewed (commented):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/gramineproject/gsc/223#-:-OJwe0OA5JuN8lorW4Br:bnfp4nl)* status: 26 of 28 files reviewed, 4 unresolved discussions, "Merge branch '" found in commit messages' one-liners (waiting on @DukeDavis12, @kailun-qin, and @woju)



<!-- Sent from Reviewable.io -->

DukeDavis12 commented:
@chiache  Please dismiss your change request. Requested change is not relevant to this PR.

DukeDavis12 commented:
@chiache Please respond.

chiache reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gsc/223)* status: 26 of 28 files reviewed, 3 unresolved discussions, "Merge branch '" found in commit messages' one-liners (waiting on @DukeDavis12, @kailun-qin, and @woju)



<!-- Sent from Reviewable.io -->

DukeDavis12 commented:
> > Reviewed 1 of 10 files at r2, 3 of 4 files at r5, 1 of 2 files at r6, 2 of 2 files at r7.
> > _[Reviewable](https://reviewable.io/reviews/gramineproject/gsc/223)_ status: 10 of 11 files reviewed, 7 unresolved discussions, not enough approvals from maintainers (2 more required), not enough approvals from different teams (2 more required, approved so far: ), one-liner is not in imperative mood (waiting on @DukeDavis12 and @woju)
> 
> _[`finalize_manifest.py` line 161 at r7](https://reviewable.io/reviews/gramineproject/gsc/223#-OF8UIxyBIWBYTYwZbFO:-OF8UIxyBIWBYTYwZbFP:bpe33qz) ([raw file](https://github.com/gramineproject/gsc/blob/1c563227497b172acbeaee205bd71fa276543838/finalize_manifest.py#L161)):_
> 
> > ```python
> >     if 'uri' not in rendered_manifest_dict['loader']['entrypoint']:
> >         rendered_manifest_dict['loader']['entrypoint']['uri'] = "file:/gramine/meson_build_output/lib/x86_64-linux-gnu/gramine/libsysdb.so"
> > ```
> 
> Why not add `/gramine/meson_build_output/lib64/gramine/runtime/glibc` to `loader.env.LD_LIBRARY_PATH` in the script?

- Removed related changes from PR.

> _[`templates/debian/entrypoint.manifest.template` line 7 at r7](https://reviewable.io/reviews/gramineproject/gsc/223#-OF8UXT20i7_Ss7Kxn9O:-OF8UXT20i7_Ss7Kxn9P:b-7ropln) ([raw file](https://github.com/gramineproject/gsc/blob/1c563227497b172acbeaee205bd71fa276543838/templates/debian/entrypoint.manifest.template#L7)):_
> 
> > ```
> > # Add "/usr/lib/x86_64-linux-gnu" explicitly because ldconfig in Ubuntu 21.04 doesn't
> > # produce it; note that this Debian template is used by Ubuntu templates as well
> > loader.env.LD_LIBRARY_PATH = "/gramine/meson_build_output/lib/x86_64-linux-gnu/gramine/runtime/glibc:/usr/lib/x86_64-linux-gnu:{{"{{library_paths}}"}}"
> > ```
> 
> ditto

- Removed related changes from PR.

> 
> _[`templates/centos/entrypoint.manifest.template` line 5 at r7](https://reviewable.io/reviews/gramineproject/gsc/223#-OF8UVXQAXgw6DnFXKeb:-OF8UVXQAXgw6DnFXKec:b-7ropln) ([raw file](https://github.com/gramineproject/gsc/blob/1c563227497b172acbeaee205bd71fa276543838/templates/centos/entrypoint.manifest.template#L5)):_
> 
> > ```
> > {% block loader %}
> > loader.env.LD_LIBRARY_PATH = "/gramine/meson_build_output/lib64/gramine/runtime/glibc:/usr/lib64:{{"{{library_paths}}"}}"
> > {% endblock %}
> > ```
> 
> ditto

- Removed related changes from PR.

DukeDavis12 commented:
> > Reviewed 1 of 1 files at r10, all commit messages.
> > _[Reviewable](https://reviewable.io/reviews/gramineproject/gsc/223)_ status: all files reviewed, 8 unresolved discussions, not enough approvals from maintainers (1 more required), not enough approvals from different teams (1 more required, approved so far: Intel) (waiting on @DukeDavis12 and @woju)
> 
> _[`-- commits` line 1 at r10](https://reviewable.io/reviews/gramineproject/gsc/223#-OG45K3n8HnjENte84B3:-OG45K3o-GJai-nVCWDv:b-e3x162):_ Pls consolidate the commits. Can be done during rebase.

- Please consolidate while merging.

kailun-qin reviewed (changes requested):
> Reviewed 1 of 2 files at r16, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gsc/223)* status: 27 of 28 files reviewed, 4 unresolved discussions (waiting on @chiache and @woju)

___
*[`-- commits` line 87 at r16](https://reviewable.io/reviews/gramineproject/gsc/223#-OKeukP10BW3R3mE8eo3:-OKeukP10BW3R3mE8eo4:b-lau9hy):*
Add `gramine-manifest-check` to GSC build



_Code quote:_
```
Add gramine-manifest-check before signing container
```
___
*[`-- commits` line 89 at r16](https://reviewable.io/reviews/gramineproject/gsc/223#-OKeuw9V-0FLIHQxN6MC:-OKeuw9W-2GkCY3u4-zA:b-9p028f):*
You don't need to replicate the title.

Instead, you should give some context and describe your changes, like: "The `gramine-manifest-check` tool for manifest schema checks was introduced in the core Gramine repository. This commit integrates it into the GSC build template and updates related manifest templates."



_Code quote:_
```
Add gramine-manifest-check before signing container.
```


<!-- Sent from Reviewable.io -->

kailun-qin reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gsc/223)* status: 27 of 28 files reviewed, 3 unresolved discussions (waiting on @chiache and @woju)

___
*[`-- commits` line 89 at r16](https://reviewable.io/reviews/gramineproject/gsc/223#-OL8JS5N7BdgF4-Zpx3P:-OL8JS5N7BdgF4-Zpx3Q:b-bcdvd2):*
Pls rewrap the commit msg to the line limit (72 chars), can be done during rebase.


<!-- Sent from Reviewable.io -->

jkr0103 reviewed (commented):
@chiache you still have two blocking comments, could you resolve?

> *[Reviewable](https://reviewable.io/reviews/gramineproject/gsc/223#-:-OLD1C3s4AOoTIY2NzyB:b-vqe3cl)* status: 27 of 28 files reviewed, 3 unresolved discussions (waiting on @chiache, @DukeDavis12, and @woju)



<!-- Sent from Reviewable.io -->

DukeDavis12 reviewed (commented):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/gramineproject/gsc/223#-:-OLcia3WB8V46C-b7xAM:bnfp4nl)* status: 27 of 28 files reviewed, 3 unresolved discussions (waiting on @chiache, @kailun-qin, and @woju)

___
*[`-- commits` line 89 at r16](https://reviewable.io/reviews/gramineproject/gsc/223#-OL8JS5N7BdgF4-Zpx3P:-OLciBkh5GL0EE3Rem6d:b-896fix):*
<details><summary><i>Previously, kailun-qin (Kailun Qin) wrote…</i></summary><blockquote>

Pls rewrap the commit msg to the line limit (72 chars), can be done during rebase.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

kailun-qin reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gsc/223)* status: 27 of 28 files reviewed, 3 unresolved discussions (waiting on @chiache, @DukeDavis12, and @woju)

___
*[`-- commits` line 89 at r16](https://reviewable.io/reviews/gramineproject/gsc/223#-OL8JS5N7BdgF4-Zpx3P:-OLdgB6a1jmgr94ZfosD:b-z0bulw):*
<details><summary><i>Previously, DukeDavis12 (Davis Benny) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Not done (rewrap to 72 chars)?


<!-- Sent from Reviewable.io -->

DukeDavis12 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gsc/223)* status: 27 of 28 files reviewed, 3 unresolved discussions (waiting on @chiache, @kailun-qin, and @woju)

___
*[`-- commits` line 89 at r16](https://reviewable.io/reviews/gramineproject/gsc/223#-OL8JS5N7BdgF4-Zpx3P:-OLmN43fCf9TUpfrUyt2:b-srcios):*
<details><summary><i>Previously, kailun-qin (Kailun Qin) wrote…</i></summary><blockquote>

Not done (rewrap to 72 chars)?
</blockquote></details>

Rewrapped to 72 Chars.


<!-- Sent from Reviewable.io -->

DukeDavis12 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gsc/223)* status: 27 of 28 files reviewed, 3 unresolved discussions (waiting on @chiache, @kailun-qin, and @woju)

___
*[`-- commits` line 89 at r16](https://reviewable.io/reviews/gramineproject/gsc/223#-OL8JS5N7BdgF4-Zpx3P:-OLmhaPSBE1el8WaAK6m:b-y75m6b):*
<details><summary><i>Previously, kailun-qin (Kailun Qin) wrote…</i></summary><blockquote>

This still looks incorrect -- why is there an indent before each line?
</blockquote></details>

Removed indent.


<!-- Sent from Reviewable.io -->

kailun-qin reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gsc/223)* status: 27 of 28 files reviewed, 2 unresolved discussions (waiting on @chiache and @woju)



<!-- Sent from Reviewable.io -->

woju reviewed (approved):
> Reviewed 2 of 2 files at r16, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gsc/223)* status: all files reviewed, 2 unresolved discussions (waiting on @chiache)



<!-- Sent from Reviewable.io -->

DukeDavis12 commented:
@chiache  please resolve your discussions asap.

chiache reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/gramineproject/gsc/223)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved



<!-- Sent from Reviewable.io -->
